druid/examples/calc.rs

```diff
@@ -112,95 +115,239 @@ impl CalcState {
+            }
+            KeyCode::Slash | KeyCode::NumpadDivide => Some(Op('÷')),
+            KeyCode::KeyC => Some(Op('c')),
+            KeyCode::Backspace => Some(Op('⌫')),
+            KeyCode::NumpadDecimal | KeyCode::Period => Some(Op('.')),
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+impl Widget<CalcState> for KeyMap {
+    fn paint(
+        &mut self,
+        paint_ctx: &mut PaintCtx,
+        base_state: &BaseState,
+        data: &CalcState,
+        env: &Env,
+    ) {
+        self.flex.paint(paint_ctx, base_state, data, env)
+    }
+
+    fn layout(
+        &mut self,
+        layout_ctx: &mut LayoutCtx,
+        bc: &BoxConstraints,
+        data: &CalcState,
+        env: &Env,
+    ) -> Size {
+        self.flex.layout(layout_ctx, bc, data, env)
+    }
+
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut CalcState, env: &Env) {
+        use crate::ButtonId::*;
+        match event {
+            Event::KeyDown(key_event) => {
+                if key_event.is_repeat {
+                    return;
+                };
+
+                let kc = key_event_to_button_id(key_event);
+                // Current we can just perform the action on the CalcState,
+                // However we don't get the pretty button animation this way.
+                match kc {
+                    None => self.flex.event(ctx, event, data, env),
+                    Some(Op(c)) => data.op(c),
+                    Some(Digit(d)) => data.digit(d),
+                }
+                match kc {
+                    None => (),
+                    Some(button_id) => {
+                        let ButtonInfo {
+                            row_id: _row_id,
+                            col_id: _col_id,
+                        } = self.key_map.get(&button_id).unwrap();
+                        // In theory if there was a mechanism to get random access into
+                        // self.flex.children we could remove the above match,
+                        // and do something fancy with the buttons here instead.
+                        ()
+                    }
+                }
+            }
+            Event::Size(_) => ctx.request_focus(),
+            _ => self.flex.event(ctx, event, data, env),
+        }
+    }
+    fn update(
+        &mut self,
+        ctx: &mut UpdateCtx,
+        old_data: Option<&CalcState>,
+        data: &CalcState,
+        env: &Env,
+    ) {
+        self.flex.update(ctx, old_data, data, env)
+    }
+}
+
+fn pad<T: Data>(inner: impl Widget<T> + 'static) -> Padding<T> {
     Padding::new(5.0, inner)
 }
 
-fn op_button_label(op: char, label: String) -> impl Widget<CalcState> {
+fn op_button_label(op: char, label: String) -> Padding<CalcState> {
     pad(Button::new(
         label,
         move |_ctx, data: &mut CalcState, _env| data.op(op),
     ))
 }
 
-fn op_button(op: char) -> impl Widget<CalcState> {
+fn op_button(op: char) -> Padding<CalcState> {
```

Comment:
Curious if this was necessary, or just a style decision?

Reply:
It was necessary, because lines [223-225](https://github.com/xi-editor/druid/pull/329/commits/3ecf4d4ce7104f903cd21fdb0d6390e6877aa2f3#diff-40cfb56c817860b90b6a65ec02961f03R223-R225)

The `impl Widget<CalcState>` erases the `Padding` type, and so the compiler could no longer distinguish that the match expressions results were of uniform type.